Sources/Moya/MoyaProvider+Stubbing.swift

```diff
@@ -0,0 +1,102 @@
+import Foundation
+
+/// Mark: Stubbing
+
+/// A TargetType configurable for testing responses
+public protocol TestTargetType {
+
+    var stubBehavior: Moya.StubBehavior { get }
+
+    var sampleResponse: EndpointSampleResponse { get }
+
+}
+
+/// Controls how stub responses are returned.
+public enum StubBehavior {
+
+    /// Do not stub.
+    case never
+
+    /// Return a response immediately.
+    case immediate
+
+    /// Return a response after a delay.
+    case delayed(seconds: TimeInterval)
+}
+
+extension MoyaProvider where Target: TestTargetType {
+
+    // swiftlint:disable function_parameter_count
+    /// When overriding this method, take care to `notifyPluginsOfImpendingStub` and to perform the stub using the `createStubFunction` method.
+    /// Note: this was previously in an extension, however it must be in the original class declaration to allow subclasses to override.
+    @discardableResult
+    open func stubRequest(_ target: Target, request: URLRequest, callbackQueue: DispatchQueue?, completion: @escaping Moya.Completion, endpoint: Endpoint<Target>, stubBehavior: Moya.StubBehavior) -> CancellableToken {
+        let callbackQueue = callbackQueue ?? self.callbackQueue
+        let cancellableToken = CancellableToken { }
+        notifyPluginsOfImpendingStub(for: request, target: target)
+        let plugins = self.plugins
+        let stub: () -> Void = createStubFunction(cancellableToken, forTarget: target, withCompletion: completion, endpoint: endpoint, plugins: plugins, request: request)
+        switch stubBehavior {
+        case .immediate:
+            switch callbackQueue {
+            case .none:
+                stub()
+            case .some(let callbackQueue):
+                callbackQueue.async(execute: stub)
+            }
+        case .delayed(let delay):
+            let killTimeOffset = Int64(CDouble(delay) * CDouble(NSEC_PER_SEC))
+            let killTime = DispatchTime.now() + Double(killTimeOffset) / Double(NSEC_PER_SEC)
+            (callbackQueue ?? DispatchQueue.main).asyncAfter(deadline: killTime) {
+                stub()
+            }
+        case .never:
+            fatalError("Method called to stub request when stubbing is disabled.")
+        }
+
+        return cancellableToken
+    }
+    // swiftlint:enable function_parameter_count
+
+    /// Creates a function which, when called, executes the appropriate stubbing behavior for the given parameters.
+    public final func createStubFunction(_ token: CancellableToken, forTarget target: Target, withCompletion completion: @escaping Moya.Completion, endpoint: Endpoint<Target>, plugins: [PluginType], request: URLRequest) -> (() -> Void) { // swiftlint:disable:this function_parameter_count
+        return {
+            if token.isCancelled {
```

Comment:
This would read better as 

```swift
guard !token.isCancelled else { 
    self.cancelCompletion(completion, target: target)
    return
}
```

IMO